src/ucp/rndv/proto_rndv.c

```diff
@@ -363,28 +346,143 @@ ucp_proto_rndv_ctrl_init(const ucp_proto_rndv_ctrl_init_params_t *params,
+    remote_select_param = ucp_proto_rndv_remote_select_param_init(params);
+    status              = ucp_proto_rndv_ctrl_select_remote_proto(
+            params, &remote_select_param, rpriv->md_map, &remote_proto_select);
+    if (status != UCS_OK) {
+        goto out;
+    }
+    remote_proto_init = &remote_proto_select->proto_init;
+
+    /* Add variants for each remote proto */
+    ucs_array_for_each(remote_proto, &remote_proto_init->protocols) {
+        remote_caps = &remote_proto->caps;
+        /* Skip empty variants and reconfig proto */
+        if ((remote_caps->num_ranges == 0) || (remote_proto->proto_id == 0)) {
+            continue;
+        }
+
+        last_range = remote_caps->ranges + (remote_caps->num_ranges - 1);
+        max_length = ucs_min(params->super.max_length, last_range->max_length);
+        min_length = ucs_max(params->super.min_length,
+                              remote_proto->caps.min_length);
+        if (max_length < min_length) {
+            continue;
+        }
+
+        remote_priv = &ucs_array_elem(&remote_proto_init->priv_buf,
+                                      remote_proto->priv_offset);
+        ucp_proto_rndv_set_variant_config(init_params, remote_proto,
+                                          &remote_select_param, remote_priv,
+                                          &rpriv->remote_proto_config);
+
+        cfg_thresh   = params->super.cfg_thresh;
+        cfg_priority = params->super.cfg_priority;
+        if ((remote_proto->cfg_thresh != UCS_MEMUNITS_AUTO)) {
+            if (cfg_thresh == UCS_MEMUNITS_AUTO) {
+                cfg_thresh = remote_proto->cfg_thresh;
+            }
+
+            /* If some remote variant has certain thresh, we need to
+             * increase their cfg_priority to force it in case of the same
+             * params->super.cfg_thresh and remote_proto->cfg_thresh.
+             *    E.g. RNDV_THRESH=0 and RNDV_MODE=get_zcopy
+             *    In this case all the variants would have 0 cfg_thresh, but
+             *    get_zcopy should be preferred over other RNDV protocols.
+             * 
+             * +1 is required in case if remote variant cfg_priority is 0.
+             */
+            if (remote_proto->cfg_thresh != UCS_MEMUNITS_INF) {
+                cfg_priority += remote_proto->cfg_priority + 1;
+            }
+        }
+
+        /* Create separate CTRL message perf node for each variant */
+        status = ucp_proto_rndv_ctrl_range_init(params, &ctrl_perf);
+        if (status != UCS_OK) {
+            break;
+        }
+
+        ucp_proto_common_init_base_caps(&params->super, min_length);
+
+        status = ucp_proto_rndv_ctrl_add_remote_variant(params, &ctrl_perf,
+                                                        remote_caps->ranges,
+                                                        min_length, max_length);
+        ucp_proto_perf_node_deref(&ctrl_perf.node);
+        if (status != UCS_OK) {
+            break;
```

Comment:
why not continue?

Reply:
As I know, fail inside `add_parallel_ranges()` can be caused only by absense of memory for allocation, so I put break here since further protos will return error in that case.

I can change it to `continue`, if it seems more logically correct from your point of view.